File: Problem102/Solution.java
import java.util.*;

import Solution.MyQueue;

public class Solution {
	public List<List<Integer>> levelOrder(TreeNode root) {
		List<List<Integer>> result = new ArrayList<>();
		MyQueue<TreeNode> queue = new MyQueue<>();
		queue.addIfNotNull(root);
		while (!queue.isEmpty()) {
			MyQueue<TreeNode> nextLevel = new MyQueue<>();
			List<Integer> currentLevel = new ArrayList<>();
			while (!queue.isEmpty()) {
				TreeNode node = queue.poll();
				currentLevel.add(node.val);
				nextLevel.addIfNotNull(node.left);
				nextLevel.addIfNotNull(node.right);
			}
			result.add(currentLevel);
			queue = nextLevel;
		}
		return result;
	}

	public class MyQueue<E> extends LinkedList<E> {
		private static final long serialVersionUID = 1L;

		public void addIfNotNull(E e) {
			if (e != null) {
				add(e);
			}
		}
	}
}

File: Problem102/Solution.go
package main

type TreeNode struct {
	Val   int
	Left  TreeNode
	Right TreeNode
}

func levelOrder(root *TreeNode) [][]int {
	var result [][]int
	if root == nil {
		return [][]int{}
	}
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		numOfNodes := len(queue)
		var level []int
		for i := 0; i < numOfNodes; i++ {
			node := queue[0]
			level = append(level, node.Val)
			queue = queue[1:]
			if node.Left != nil {
				queue = append(queue, node.Left)
			}

			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		result = append(result, level)
	}
	return result
}
